tools/outline_proxy_controller/network_monitor.cpp

```diff
@@ -0,0 +1,148 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <array>
+#include <cstddef>      // size_t, byte
+#include <cstring>      // memset
+#include <system_error>
+#include <type_traits>  // underlying_type_t, is_integral
+
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+
+#include <boost/asio.hpp>
+
+#include "logger.h"
+#include "network_monitor.h"
+
+using namespace outline;
+
+static const std::size_t kRecvBufferSize = 4096;
+
+NetworkMonitor::NetworkMonitor(const boost::asio::any_io_executor &io_context)
+  : netlink_socket_{io_context}
+{
+  using namespace boost::asio::generic;
+
+  sockaddr_nl sa;
+  std::memset(&sa, 0, sizeof(sa));
+  sa.nl_family = AF_NETLINK;
+  sa.nl_groups = RTMGRP_LINK         // network card change
+               | RTMGRP_IPV4_IFADDR  // IPv4 address change
+               | RTMGRP_IPV4_ROUTE   // IPv4 route table change
+               | RTMGRP_IPV6_IFADDR  // IPv6 address change
+               | RTMGRP_IPV6_ROUTE;  // IPv6 route table change
+
+  netlink_socket_.open({AF_NETLINK, NETLINK_ROUTE});
+  netlink_socket_.bind({&sa, sizeof(sa)});
+
+  // for non-blocking peek IsNextChangeEventReady
+  netlink_socket_.non_blocking(true);
+
+  logger.info("network monitor initialized");
+}
+
+NetworkMonitor::~NetworkMonitor() {
+  logger.info("network monitor destroyed");
+}
+
+boost::asio::awaitable<NetworkChangeEvent> NetworkMonitor::WaitForChangeEvent() {
+  using namespace boost::asio;
+
+  std::array<std::byte, kRecvBufferSize> buf;
+  NetworkChangeEvent received_events;
+
+  do {
+    auto len = co_await netlink_socket_.async_receive(buffer(buf), use_awaitable);
+
+    // https://linux.die.net/man/7/netlink
+    // https://man7.org/linux/man-pages/man7/rtnetlink.7.html
+    for (auto msg = reinterpret_cast<nlmsghdr*>(buf.data()); NLMSG_OK(msg, len); msg = NLMSG_NEXT(msg, len)) {
+      switch (msg->nlmsg_type) {
+        case NLMSG_DONE:
+          goto EndOfMultiPartMsg;
+        case NLMSG_ERROR:
+        {
+          auto err = reinterpret_cast<nlmsgerr*>(NLMSG_DATA(msg));
+          throw std::system_error{err->error, std::system_category()};
+        }
+        case RTM_NEWLINK:
+          received_events.SetNicChanged();
+          break;
+        case RTM_NEWADDR:
+        case RTM_DELADDR:
+          received_events.SetAddressChanged();
+          break;
+        case RTM_NEWROUTE:
+        case RTM_DELROUTE:
+          received_events.SetRoutingChanged();
+          break;
+      }
+    }
+
+EndOfMultiPartMsg:
+    ;
+  } while (received_events.IsEmpty() || IsNextChangeEventReady());
```

Comment:
Added `IsNextChangeEventReady` here because netlink events are pretty noisy, for example, our configuration will lead to ~10 change events, but they are immediately ready at once. So here I try to aggregate them all and just return once (with the event type bits set).

Reply:
Please add a comment explaining that. We will forget.

FYI, we had a similar issue on iOS of too many network change notifications: https://github.com/Jigsaw-Code/outline-client/blob/009db3b984d00c2740d6db1b5217c5abc2428574/cordova-plugin-outline/apple/vpn/PacketTunnelProvider.m#L358

Reply:
Good to know, and Windows has similar issue as well: https://github.com/Jigsaw-Code/outline-client/blob/master/tools/OutlineService/OutlineService/OutlineService.cs#L908-L912

Reply:
I still don't see the comment in the code explaining why we need that.

Reply:
updated.